udpsession.go

```diff
@@ -0,0 +1,658 @@
+	}
+
+	for _, usa := range us.sessKeyToSessAddr { // Send close message to all dialers.
+		if tcpConn, ok := us.tcpConns[usa.String()]; ok {
+			_, err = tcpConn.Write(b)
+		}
+	}
+
+	return
+}
+
+// Handle connection closed by peer or getting closed message from peer.
+func (us *UdpSession) handleConnClosed(sessKey string) {
+	us.Lock()
+	defer us.Unlock()
+	if us.isListener {
+		if usa, ok := us.sessKeyToSessAddr[sessKey]; ok {
+			delete(us.sessKeyToSessAddr, sessKey)
+			delete(us.udpAddrToSessAddr, usa.udpAddr.String())
+			if tcpConn, ok := us.tcpConns[sessKey]; ok {
+				tcpConn.Close()
+				delete(us.tcpConns, sessKey)
+			}
+		}
+	} else {
+		us.peer = nil
+		us.udpConn.Close()
+		us.tcpConn.Close()
+	}
+}
+
+// Get UDPAddr by net.Addr
+func (us *UdpSession) getUdpAddr(addr net.Addr) (udpAddr *net.UDPAddr, err error) {
+	us.RLock()
+	defer us.RUnlock()
+
+	if !us.isListener {
+		udpAddr = us.peer.udpAddr
+		return
+	}
+	// for listener
+	if addr == nil {
+		return nil, ErrWrongAddr
+	} else {
+		usa, ok := addr.(*udpSessionAddr) // addr should be *udpSessionAddr
+		if !ok {
+			return nil, ErrWrongAddr
+		}
+
+		usa, ok = us.sessKeyToSessAddr[usa.String()]
+		if ok {
+			udpAddr = usa.udpAddr
+		} else {
+			return nil, ErrWrongAddr
+		}
+	}
+
+	return
+}
+
+func (us *UdpSession) sendMetaData(sessionID []byte) (err error) {
+	metadata := &pb.SessionMetadata{
+		Id:            sessionID,
+		SessionType:   pb.SessionType_UDP,
+		DialerNknAddr: us.ts.Address(),
+	}
+	buf, err := proto.Marshal(metadata)
+	if err != nil {
+		return err
+	}
+
+	b := newSessionData(pb.MsgType_SESSIONID, buf)
+
+	us.RLock()
+	_, _, err = us.udpConn.WriteMsgUDP(b, nil, nil)
+	us.RUnlock()
+
+	return err
+}
+
+// Handle session meta data.
+func (us *UdpSession) handleMetaData(b []byte, udpAddr *net.UDPAddr) error {
+	if !us.isListener { // dialer should not get session meta data.
+		return nil
+	}
+
+	sessData, err := parseSessionData(b)
+	if err != nil {
+		return err
+	}
+
+	switch sessData.MsgType {
+	case pb.MsgType_SESSIONID:
+		metadata := &pb.SessionMetadata{}
+		err = proto.Unmarshal(sessData.Data, metadata)
+		if err != nil {
+			return err
+		}
+		usa := &udpSessionAddr{sessionID: metadata.Id, remoteNknAddr: metadata.DialerNknAddr, udpAddr: udpAddr}
+		if _, ok := us.sessKeyToSessAddr[usa.String()]; ok { // This session is exist already
+			return nil
+		}
+
+		us.Lock()
+		us.sessKeyToSessAddr[usa.String()] = usa
+		us.udpAddrToSessAddr[udpAddr.String()] = usa
+		us.Unlock()
+		us.addCodec(udpAddr, metadata.DialerNknAddr) // listener add codec
+
+	case pb.MsgType_CLOSE:
+		if usa, ok := us.udpAddrToSessAddr[udpAddr.String()]; ok {
+			us.handleConnClosed(usa.String())
+		}
+
+	default:
+		return ErrWrongMsgFormat
+	}
+
+	return nil
+}
+
+// New session data
+func newSessionData(msgType pb.MsgType, data []byte) (b []byte) {
+	sessData := pb.SessionData{MsgType: msgType, Data: data}
+	buf, _ := proto.Marshal(&sessData)
+	b = make([]byte, len(buf)+1)

```

Comment:
A shorter version:
```
b = append([]byte{pb.HeaderType_SESSION}, buf...)
```

Reply:
Yes, modified. But I just wonder why golang designer doesn't use append([]a, []b), but uses append([]a, []b...), will it append one by one from b? Will it be slower than copy(a, b)? if append one by one, slice a need allocate new buffer very often.